implementations/rust/ockam/ockam_ffi/src/vault.rs

```diff
@@ -423,53 +420,96 @@ pub extern "C" fn ockam_vault_aead_aes_gcm_decrypt(
-    };
+        let ciphertext_and_tag = unsafe {
+            core::slice::from_raw_parts(ciphertext_and_tag, ciphertext_and_tag_length as usize)
+        };
 
-    match block_future(async move {
-        let mut v = get_vault(context).await?;
-        let ctx = Secret::new(secret as usize);
-        let mut nonce_vec = vec![0; 12 - 2];
-        nonce_vec.extend_from_slice(&nonce.to_be_bytes());
-        let plain = v
-            .aead_aes_gcm_decrypt(&ctx, ciphertext_and_tag, &nonce_vec, additional_data)
-            .await?;
-        if plaintext_size < plain.len() as u32 {
-            return Err(FfiError::BufferTooSmall.into());
-        }
-        *plaintext_length = plain.len() as u32;
+        block_future(async move {
+            let mut v = get_vault(context).await?;
+            let ctx = Secret::new(secret as usize);
+            let mut nonce_vec = vec![0; 12 - 2];
+            nonce_vec.extend_from_slice(&nonce.to_be_bytes());
+            let plain = v
+                .aead_aes_gcm_decrypt(&ctx, ciphertext_and_tag, &nonce_vec, additional_data)
+                .await?;
+            if plaintext_size < plain.len() as u32 {
+                return Err(FfiError::BufferTooSmall.into());
+            }
+            *plaintext_length = plain.len() as u32;
 
-        unsafe { std::ptr::copy_nonoverlapping(plain.as_ptr(), plaintext, plain.len()) };
-        Ok::<(), Error>(())
-    }) {
-        Ok(_) => FfiOckamError::none(),
-        Err(err) => err.into(),
-    }
+            unsafe { std::ptr::copy_nonoverlapping(plain.as_ptr(), plaintext, plain.len()) };
+            Ok::<(), Error>(())
+        })?;
+        Ok(())
+    })
 }
 
 /// De-initialize an Ockam Vault.
 #[no_mangle]
 pub extern "C" fn ockam_vault_deinit(context: FfiVaultFatPointer) -> FfiOckamError {
-    block_future(async move {
-        match context.vault_type() {
-            FfiVaultType::Software => {
-                let handle = context.handle() as usize;
-                let mut v = SOFTWARE_VAULTS.write().await;
-                if handle < v.len() {
-                    v.remove(handle);
-                    FfiOckamError::none()
-                } else {
-                    FfiError::VaultNotFound.into()
+    handle_panics(|| {
+        block_future(async move {
+            match context.vault_type() {
+                FfiVaultType::Software => {
+                    let handle = context.handle() as usize;
+                    let mut v = SOFTWARE_VAULTS.write().await;
+                    if handle < v.len() {
+                        v.remove(handle);
+                        Ok(())
+                    } else {
+                        Err(FfiError::VaultNotFound)
+                    }
                 }
             }
-        }
+        })?;
+        Ok(())
     })
 }
+
+fn handle_panics<F>(f: F) -> FfiOckamError
+where
+    F: FnOnce() -> Result<(), FfiOckamError>,
+{
+    let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(f));
```

Comment:
AssertUnwindSafe here doesn't really matter — `UnwindSafe` is dubious at best, is unrelated to safety, and we aren't in a position to properly represent it anyway (as propagating the unwinding would be UB).